src/main/scala/org/constellation/primitives/TrieBasedTipService.scala

```diff
@@ -1,32 +1,29 @@
 package org.constellation.primitives
+import cats.effect.IO
 import com.typesafe.scalalogging.Logger
 import org.constellation.DAO
 import org.constellation.consensus.TipData
-import org.constellation.primitives.Schema.{GenesisObservation, Id, SignedObservationEdge}
+import org.constellation.primitives.Schema.{Id, SignedObservationEdge}
 import org.constellation.util.Metrics
 
 import scala.collection.concurrent.TrieMap
-import scala.util.Random
+import scala.util.{Failure, Random, Success, Try}
 
 trait ConcurrentTipService {
 
   def getMinTipHeight()(implicit dao: DAO): Long
   def toMap: Map[String, TipData]
-  def toSeq: Seq[CheckpointBlock]
   def size: Int
   def set(tips: Map[String, TipData])
-  def update(checkpointBlock: CheckpointBlock)(implicit dao: DAO)
-  // considerd as private only
-  def put(go: GenesisObservation)(implicit metrics: Metrics): Option[TipData]
+  def update(checkpointBlock: CheckpointBlock)(implicit dao: DAO): IO[Try[Option[TipData]]]
```

Comment:
so `IO` or `Try`? IMO it shouldn't be mixed

Reply:
Well there are 3 states to be handled:
- Tip was inserted
- Tip was updated
- Tip was conflicting

I could use IO and throw ex when necessary, and move handle logic to upper levels but then I would need to do same several times :thinking:

Reply:
For me inserted/updated is the same if you're using `update`. If you would like to insert, you should check if the element is present or not and then put if necessary.
Anyway, I'd see it like: `IO[Either[TipUpdateError, TipData]]`.

* IO[Right[TipData]] - tip was inserted or updated
* IO[Left[TipConflicting]] - tip was conflicting